Add partial-type-name suggestions to C# and some fixes for VB

lorcanmooney opened the pull request:
- Closes #7213. Refactors the VB provider, creating a base class and a C# implementation.
- Fixes #4137. In VB, when a user commits a generic type using `(` only the type name will be inserted.
- Fixes an issue in the VB provider where suggestions we pulled from all ancestor types and namespaces, instead of just the directly enclosing container.
- Fixes an issue in the VB provider where types wouldn't be suggested because their spans overlapped, despite being in different documents.

lorcanmooney commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -679,6 +679,9 @@ Microsoft.CodeAnalysis.SourceGeneratorAttribute.SourceGeneratorAttribute(string
 Microsoft.CodeAnalysis.SourceGeneratorExtensions
+Microsoft.CodeAnalysis.SymbolDisplayFormat.RemoveGenericsOptions(Microsoft.CodeAnalysis.SymbolDisplayGenericsOptions options) -> Microsoft.CodeAnalysis.SymbolDisplayFormat
+Microsoft.CodeAnalysis.SymbolDisplayFormat.RemoveLocalOptions(Microsoft.CodeAnalysis.SymbolDisplayLocalOptions options) -> Microsoft.CodeAnalysis.SymbolDisplayFormat
+Microsoft.CodeAnalysis.SymbolDisplayFormat.RemoveMiscellaneousOptions(Microsoft.CodeAnalysis.SymbolDisplayMiscellaneousOptions options) -> Microsoft.CodeAnalysis.SymbolDisplayFormat
```
Only `RemoveMiscellaneousOptions` is used by this change. The others seemed like omissions based on the existing API.

I don't know what the process is for public API changes, but if necessary I can pull these out and rework the code so it's not needed.

CyrusNajmabadi replied:
This looks like an unintentional omission as we already have "RemoveKindOptions" "RemoveParameterOptions" etc..  Tagging @dotnet/roslyn-compiler to sign off on this bit.  Personally i think it is appropriate.

lorcanmooney commented on src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb:
```diff
@@ -864,7 +864,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Private Function CheckSymbolLocationsAgainstSyntax(symbol As NamedTypeSymbol, nodeToCheck As VisualBasicSyntaxNode) As NamedTypeSymbol
             For Each location In symbol.Locations
-                If location.SourceTree Is Me.SyntaxTree AndAlso nodeToCheck.Span.Contains(location.SourceSpan.Start) Then
+                If location.SourceTree Is Me.SyntaxTree AndAlso nodeToCheck.Span.Contains(location.SourceSpan) Then
```
This makes the behaviour (and code) of `GetDeclaredSymbol` consistent with the C# semantic model. Without this, the VB implementation would return `Nothing` for types if the user hasn't typed a name yet.

Pilchie replied:
Tagging @AlekseyTs - what do you think about this change?

CyrusNajmabadi replied:
Consistency would definitely be appreciated here.  @lorcanmooney have you written a test for this new behavior in the compiler layer?  If not, @AlekseyTs can you point lorcan to the right place to add such a test?

lorcanmooney replied:
I haven't added a test yet, and there don't seem to be _any_ tests for this, at least not that I could find. If somebody can point me in the right direction, I'll gladly add a few.

CyrusNajmabadi replied:
I think teh right place would probably be: Open\src\Compilers\VisualBasic\Test\Semantic\Compilation\SemanticModelGetDeclaredSymbolAPITests.vb

Since you're doing this to affect how GetDeclaredSymbol works right?

lorcanmooney replied:
Yeah, that looks like the place. I try to fill these out tomorrow.

AlekseyTs replied:
> Without this, the VB implementation would return `Nothing` for types if the user hasn't typed a name yet.

I am probably missing something obvious, but it looks like the condition became tighter. So, I am not sure how it makes the function to return Nothing less often. I think, first a problem should be demonstrated and only then addressed. Without that, I wouldn't keep the change.     

---
In reply to: [67624193](https://github.com/dotnet/roslyn/pull/12092#discussion_r67624193) [](ancestors = 67624193)

lorcanmooney replied:
Imagine a user creating a new partial class:
```
Partial Class $$
```
In the above, `location.SourceSpan.Start` will be 14 (the start of the missing identifier). As it is now, the call to `Contains(int)` treats the node span as a right-open interval [0, 14), so the check fails. Changing the code to use `Contains(Span)` - the span being [14,14) - returns `True` for this case.

AlekseyTs replied:
Please add a GetDeclaredSymbol unit-test for this scenario that fails without this change and succeeds with it.

lorcanmooney replied:
Done. Added `TestGetDeclaredSymbolWithIncompleteDeclaration` to SemanticModelGetDeclaredSymbolAPITests.vb.

jcouv replied:
Thanks for the VB test. Would you be kind to also add the corresponding C# test?

LGTM aside from this and the question on public APIs.

lorcanmooney replied:
Good point. I'll add the test.

lorcanmooney commented on src/EditorFeatures/VisualBasicTest/Diagnostics/GenerateConstructor/GenerateConstructorTests.vb:
```diff
@@ -566,8 +566,8 @@ End Class")
-        Public Async Function TestGenerateInDerivedType_Crash() As Task
-            Await TestMissingAsync(
+        Public Async Function TestGenerateInDerivedType_InvalidClassStatement() As Task
+            Await TestAsync(
```
The change to the VB semantic model means this scenario works properly instead of having to work around a crash.

Pilchie commented:
Tag @roslyn-ide for review,

@dotnet-bot test eta please, test vsi please, retest  perf_correctness_prtest please.

Pilchie commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PartialTypeCompletionProviderTests.cs:
```diff
@@ -0,0 +1,291 @@
+
+partial class C<T> ";
+
+            await VerifyProviderCommitAsync(text, "C<T>", expected, ' ', "");
```
This is what I would expect if the item was explicitly selected - should we also have a test that verifies the builder is showing so that typing a new name that is a substring of an existing partial class doesn't complete to the existing thing?

Also, what about committing a generic with <?  I would expect just `C<` to be inserted in that case?

Pilchie replied:
And, now I see the suggestion mode provider tests ;)

lorcanmooney replied:
> Also, what about committing a generic with <? I would expect just  C<  to be inserted in that case?

I expected that too, but I couldn't find any precedent for that behaviour. I used the `ObjectCreationCompletionProvider` as my reference point for this. In VB, it also special-cases insertions via `(`, but in C# that provider doesn't commit _anything_ if the user presses `<`, it simply closes the completion list.

So, an easy fix, but on it's own will introduce some inconsistency between the various providers. Maybe `ObjectCreationCompletionProvider` should be updated to honour `<` as well?

Pilchie replied:
I think the background here is that the C# symbol provider doesn't ever commit the `<`.  This is because it never really knows any instantiation info, so it can't commit anything useful except the `<` itself, which is just one character and so didn't warrant it.

Providers that *do* know instantiation info (this, `ObjectCreationCompletionProvider`, the cref provider, etc) don't tend to handle the overtyping case well in C#.  We could definitely change that, but I don't think it needs to be part of this PR.

CyrusNajmabadi replied:
I think this new behavior is ok.  Because it's not explicitly selected, the user has to go and select that item explicitly.  Selecting it explicitly and then typing '<' is somewhat strange as you'd be effectively saying "i want completion's help in typing the type name... except for the generics, which i will type myself"

I more expect people to go and select the item in the list and then type "space, tab, enter or colon".  All of which will have the right behavior.  So i think how this is working now is ok.  What do you think Kevin?

Pilchie commented on src/Features/VisualBasic/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.vb:
```diff
@@ -1,150 +1,68 @@
-                Dim displayAndInsertionText = CompletionUtilities.GetDisplayAndInsertionText(symbol, isAttributeNameContext:=False, isAfterDot:=False, isWithinAsyncMethod:=False, syntaxFacts:=context.GetLanguageService(Of ISyntaxFactsService))
-                displayText = displayAndInsertionText.Item1
-                insertionText = displayAndInsertionText.Item2
+            If ch = "("c Then
```
It seems odd that we have to do this in this way.

lorcanmooney replied:
So, this one has a bit of history. In short, I tried inheriting from `AbstractSymbolCompletionProvider` which would handle this, but it seemed to overcomplicate things. In the end, this is pretty much [how it's done by the other providers](https://github.com/dotnet/roslyn/blob/4333bf16afc8b94b1d25ef3d9dbf4e125f16a6d2/src/Features/VisualBasic/Portable/Completion/CompletionProviders/CompletionUtilities.vb#L129), just more up-front.

lorcanmooney replied:
It seems using `AbstractSymbolCompletionProvider` would also change the commit-on-space behaviour so, one way or another, we're going to end up special-casing something here.

Pilchie replied:
Thanks for the pointer - this looks okay to me now.  Again, I think the difference comes down to the fact that the `SymbolCompletionProvider` doesn't know the generic instantiation info, so the user is always going to have to fill it in.

Pilchie commented:
@CyrusNajmabadi @rchande - can you take a look at the way that typing `(` in a VB generic type is handled to not overtype?

Pilchie commented:
Generally, I'm 👍 on this - but would like compiler signoff on the symbol display and semantic model changes, and word from either @rchande or @CyrusNajmabadi about the right way to handle changing the insertion text.

lorcanmooney commented:
@Pilchie Thanks for the review and background info. It helps to have some context for _why_ things are the way they are. I've pushed an update to special-case `<` in the new C# provider.

I'd like to try to fix #4585 next. It will require some minor refactoring of the generic-aware providers, but I have some ideas and I think I'll be able to remove the duplicated bits of special-casing introduced here. It will also cross into the `ObjectCreationCompletionProvider` so if you'd like the `<` commit fixed or other tweaks in the area, please let me know.

lorcanmooney commented:
> error VSSDK1031: Extension '49e24138-9ee3-49e0-8ede-6b39f49303bf' could not be found. Please make sure the extension has been installed.

@dotnet-bot retest perf_correctness_prtest

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeaturesResources.resx:
```diff
@@ -461,4 +461,22 @@
   <data name="NamespaceName" xml:space="preserve">
     <value>&lt;namespace name&gt;</value>
   </data>
+  <data name="AutoselectDisabledDueToClassDeclaration" xml:space="preserve">
```
I would just have one of these and say "due to Type declaration."

lorcanmooney replied:
I actually had that initially, but then I realised I was just being lazy :-) Unless there's localisation overhead you want to avoid, I'd like to leave it as is.

CyrusNajmabadi replied:
So, IMO, i prefer to eschew excess code that has little marginal utility.  In this case it seems to be excess code that doesn't actually end up improving the user experience in any real way.  

It's similar to if we were to take every error that might be put on an an expression and were to specialize it for every type of expression.

I like localization specialization where there is something practically different that we'd be saying.  Honestly, for the entire class of messages that are nothing more than "completion is in builder mode because you could be declaring a variable" i would be fine with a single message.  I don't think there's value actually saying "you could be declaring a namespace" or a struct, or a class.  It's all clear from context.  About the only one that is slightly special is lambdas.  That's because there is no clear syntax when you're writing it that you're making a lambda.  And it can definitely be non-obvious why we picked the builder.

So, personally, i'd prefer this code to be removed as i don't see it justifying itself.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.cs:
```diff
@@ -0,0 +1,96 @@
+                typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+                genericsOptions:
+                    SymbolDisplayGenericsOptions.IncludeTypeParameters |
+                    SymbolDisplayGenericsOptions.IncludeVariance,
```
Do you have a test for variance?

lorcanmooney replied:
Yup. Last test in PartialTypeCompletionProviderTests.cs

CyrusNajmabadi replied:
Great!

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.cs:
```diff
@@ -0,0 +1,96 @@
+        internal override bool IsInsertionTrigger(SourceText text, int characterPosition, OptionSet options)
+        {
+            var ch = text[characterPosition];
+            return ch == ' ' || (CompletionUtilities.IsStartingNewWord(text, characterPosition) && options.GetOption(CompletionOptions.TriggerOnTypingLetters, LanguageNames.CSharp));
```
please wrap lines that are too long to fit in a github PR :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.cs:
```diff
@@ -0,0 +1,96 @@
+                {
+                    var position = SymbolCompletionItem.GetContextPosition(selectedItem);
+                    var semanticModel = await document.GetSemanticModelForSpanAsync(new TextSpan(position, 0), cancellationToken).ConfigureAwait(false);
+                    var insertionText = symbols[0].ToMinimalDisplayString(semanticModel, position, _symbolFormatWithoutGenerics);
```
Wouldn't it be fairly simple to just use symbol.Name here and avoid all the other complexity?

lorcanmooney replied:
This also deals with escaping things like `@class`. I have an idea that should reduce a bunch of this stuff as part of #4585, so I'll come back to this later.

CyrusNajmabadi replied:
> This also deals with escaping things like @class

That's pretty easily handled by just saying: symbol.Name.EscapeIdentifier() though.  SymbolDisplay is really a good choice when it's a long and complicated name (i.e. when there's a signature or generics in it).  But when there's literally just the name alone, it serves little purpose.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/SuggestionMode/CSharpSuggestionModeCompletionProvider.cs:
```diff
@@ -63,6 +63,20 @@ internal class CSharpSuggestionModeCompletionProvider : SuggestionModeCompletion
+                    switch (typeDeclaration.Keyword.Kind())
+                    {
+                        case SyntaxKind.ClassKeyword:
+                            return CreateSuggestionModeItem(CSharpFeaturesResources.ClassName, itemSpan, CSharpFeaturesResources.AutoselectDisabledDueToClassDeclaration);
```
I don't think this level of specificity is necessary.

lorcanmooney replied:
As above, I think it's nice to keep attention to detail like this. That said, I don't feel strongly about it, so if there's some hidden cost, I happily trim it back.

CyrusNajmabadi replied:
If you prefer it, then let's go with this. :)

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs:
```diff
@@ -27,11 +27,11 @@ internal abstract partial class AbstractPartialCompletionProvider : AbstractMemb
                         SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
 
-        public AbstractPartialCompletionProvider()
+        public AbstractPartialMethodCompletionProvider()
```
nit: make protected.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.vb:
```diff
@@ -1,150 +1,67 @@
+        Inherits AbstractPartialTypeCompletionProvider
 
-        Private ReadOnly _partialNameFormat As SymbolDisplayFormat =
+        Private Shared ReadOnly _insertionTextFormatWithGenerics As SymbolDisplayFormat =
```
Is this format different from teh C# one?  If not, pull up to base class.

lorcanmooney replied:
This one includes generic constraints.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Completion/SuggestionMode/VisualBasicSuggestionModeCompletionProvider.vb:
```diff
@@ -97,11 +97,45 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.SuggestionMode
 
                 description = VBFeaturesResources.TypeANameHereToDeclareANamespace & vbCrLf &
-                              VBFeaturesResources.NoteSpaceAndCompletion
+                              VBFeaturesResources.NoteSpaceCompletionIsDisa
```
"NoteSpaceCompletionIsDisa"??

lorcanmooney replied:
Yeah, what's up with that? ;-) Both of these are existing resources. As much as I want to fix it, I suspect changing the name might interfere with the loc process.

As for the change itself, the new one has better wording - it doesn't mention anything about committing on `=`.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -618,6 +618,33 @@ public static bool IsNamespaceDeclarationNameContext(this SyntaxTree syntaxTree,
+            }
+
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken)
+                            .GetPreviousTokenIfTouchingWord(position);
```
align dots.

lorcanmooney replied:
Argh! I can tell how many time's I've had to fix that!

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -618,6 +618,33 @@ public static bool IsNamespaceDeclarationNameContext(this SyntaxTree syntaxTree,
+                    declarationSyntax = token.GetAncestor<TypeDeclarationSyntax>();
+                    return declarationSyntax != null &&
+                           declarationSyntax.Keyword == token &&
+                           declarationSyntax.Modifiers.Any(SyntaxKind.PartialKeyword);
```
it's actually much simpler in C#.  "partial" has to immediately precede class/interface/struct.  So you can just walk back one token and check that.

lorcanmooney replied:
By using `GetPreviousToken`? I'm not sure what the `includeZeroWidth` and  `includeSkipped` refer to. Can I just accept the defaults?

CyrusNajmabadi replied:
Yup.  Just use the normal ```GetPreviousToken()``` with no args.  The args are for some *very* specialized error recovery behavior.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -618,6 +618,33 @@ public static bool IsNamespaceDeclarationNameContext(this SyntaxTree syntaxTree,
+                           declarationSyntax.Modifiers.Any(SyntaxKind.PartialKeyword);
+
+                default:
+                    declarationSyntax = null;
```
duplciated with the top of the method.  My preferred pattern would be:

```c#
if (!IsInNonUser...)
{ 
     ...
}

failure case
```

lorcanmooney replied:
Will fix.

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.vb:
```diff
@@ -75,6 +75,30 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions.ContextQuery
+                    statementSyntax = token.GetAncestor(Of TypeStatementSyntax)()
+                    Return statementSyntax IsNot Nothing AndAlso
+                           statementSyntax.DeclarationKeyword = token AndAlso
+                           statementSyntax.Modifiers.Any(SyntaxKind.PartialKeyword)
```
note: my comment about 'partial' only applies to C#.  For VB you do need this code.

CyrusNajmabadi commented:
Looking very close.  Thanks @lorcanmooney !

lorcanmooney commented on src/Features/CSharp/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.cs:
```diff
@@ -0,0 +1,96 @@
+        private static readonly SymbolDisplayFormat _symbolFormatWithGenerics =
+            new SymbolDisplayFormat(
+                globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Omitted,
+                typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
```
Yes, for C# I think you're right. The VB side may include the type constraints which could need qualification.

CyrusNajmabadi replied:
Interesting point about the constraints.  Do you have tests that ensure htat we don't spit out a qualified name though from the type name itself?  That will never be right in either language.

As for the constraints, does VB work like C#?  If so, you could avoid putting the constraints in in the partial type?  That's how we generally write partial classes in C# so we avoid a lot of repetition of things like the base types, constraints, etc.

lorcanmooney replied:
> As for the constraints, does VB work like C#?

Doesn't seem so. I get a compiler error unless all the declarations repeat the same set of constraints.

> Do you have tests that ensure htat we don't spit out a qualified name though from the type name itself?

We do (see `TestPartialClassWithSameMemberName`). But we also have a bug (#11569) because with the right naming, it already breaks. I don't fully understand it yet, but I think it's because the insertion position is considered _inside_ the type (another VB/C# semantic difference as far as I can tell).

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs:
```diff
@@ -27,11 +27,11 @@ internal abstract partial class AbstractPartialCompletionProvider : AbstractMemb
-        protected abstract bool IsPartialCompletionContext(SyntaxTree tree, int position, CancellationToken cancellationToken, out DeclarationModifiers modifiers, out SyntaxToken token);
+        protected abstract bool IsPartialMethodCompletionContext(SyntaxTree tree, int position, CancellationToken cancellationToken, out DeclarationModifiers modifiers, out SyntaxToken token);
         protected abstract string GetDisplayText(IMethodSymbol method, SemanticModel semanticModel, int position);
         protected abstract bool IsPartial(IMethodSymbol m);
```
not part of your change.  but if you could rename this from 'm' to 'method' that would be appreciated.

lorcanmooney replied:
Sure.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/PartialTypeCompletionProvider.cs:
```diff
@@ -0,0 +1,95 @@
+        private static readonly SymbolDisplayFormat _symbolFormatWithoutGenerics =
+            _symbolFormatWithGenerics.WithGenericsOptions(SymbolDisplayGenericsOptions.None);
+
+        public PartialTypeCompletionProvider()
```
not needed.

lorcanmooney replied:
Gone.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractPartialTypeCompletionProvider.cs:
```diff
@@ -0,0 +1,120 @@
+
+        protected abstract ValueTuple<string, string> GetDisplayAndInsertionText(INamedTypeSymbol symbol, AbstractSyntaxContext context);
+
+        protected virtual IEnumerable<INamedTypeSymbol> LookupCandidateSymbols(AbstractSyntaxContext context, INamedTypeSymbol declaredSymbol, CancellationToken cancellationToken)
```
Does this need to be virtual?

lorcanmooney replied:
Yes, the C# implementation needs to refine the results to only those with a `partial` modifier.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -618,6 +618,29 @@ public static bool IsNamespaceDeclarationNameContext(this SyntaxTree syntaxTree,
+                    case SyntaxKind.ClassKeyword:
+                    case SyntaxKind.StructKeyword:
+                    case SyntaxKind.InterfaceKeyword:
+                        declarationSyntax = token.GetAncestor<TypeDeclarationSyntax>();
```
don't even need to get declaration syntax.  Literally just get the previous token and see if it is "partial" :)

lorcanmooney replied:
I've reworked it so the declaration is only retrieved if both token tests pass, but the declaration is still needed by the caller.

CyrusNajmabadi commented:
tiny nits.  LGTM

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Compilation/SemanticModelGetDeclaredSymbolAPITests.vb:
```diff
@@ -2933,6 +2933,23 @@ End Interface
+            Dim syntax = DirectCast(tree.FindNodeOrTokenByKind(SyntaxKind.ClassStatement, 1).AsNode(), ClassStatementSyntax)
+            Dim symbol = model.GetDeclaredSymbol(syntax)
+            Assert.NotNull(symbol)
+            Assert.Equal("?", symbol.ToTestDisplayString())
```
In addition to this check, consider asserting that the symbol represents a class.

lorcanmooney replied:
Done. I've also added a couple of surrounding declarations to make sure the _right_ symbol is returned.

AlekseyTs commented:
Changes under in compilers LGTM, modulo a small suggestion for the TestGetDeclaredSymbolWithIncompleteDeclaration unit-test. Tagging @dotnet/roslyn-compiler for the second sign-off.

AlekseyTs commented:
LGTM

lorcanmooney commented:
I've rebased, squashed and resolved fixed a minor merge conflict.

This still needs a final sign-off from @dotnet/roslyn-compiler. Anyone feel like taking a look?

jcouv commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PartialTypeCompletionProviderTests.cs:
```diff
@@ -0,0 +1,311 @@
+            return new PartialTypeCompletionProvider();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
```
Consider moving the traits onto the class.

jcouv commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -671,6 +671,9 @@ Microsoft.CodeAnalysis.Semantics.UnaryOperationKind.UnsignedPostfixDecrement = 7
 Microsoft.CodeAnalysis.Semantics.UnaryOperationKind.UnsignedPrefixIncrement = 773 -> Microsoft.CodeAnalysis.Semantics.UnaryOperationKind
+Microsoft.CodeAnalysis.SymbolDisplayFormat.RemoveGenericsOptions(Microsoft.CodeAnalysis.SymbolDisplayGenericsOptions options) -> Microsoft.CodeAnalysis.SymbolDisplayFormat
+Microsoft.CodeAnalysis.SymbolDisplayFormat.RemoveLocalOptions(Microsoft.CodeAnalysis.SymbolDisplayLocalOptions options) -> Microsoft.CodeAnalysis.SymbolDisplayFormat
+Microsoft.CodeAnalysis.SymbolDisplayFormat.RemoveMiscellaneousOptions(Microsoft.CodeAnalysis.SymbolDisplayMiscellaneousOptions options) -> Microsoft.CodeAnalysis.SymbolDisplayFormat
```
It seems that, out of the three new public methods, only `RemoveMiscellaneousOptions` is used in the code, the other two new APIs are un-used at this point. Are there here just for consistency?

Also, all three Remove APIs are just `AddSomeOption(~option)`. If we're going to be consistent, are we going to have Remove APIs whenever we have Add APIs?

Was there a discussion on whether we need those new public APIs?

lorcanmooney replied:
> Are there here just for consistency?

Yes.

> If we're going to be consistent, are we going to have Remove APIs whenever we have Add APIs?

It seemed like that was already the intention since there were already a handful of `Remove*` methods. Since I was adding one more, I thought adding the last two would be the right thing to do. Taking a broader view, it seems odd to have full set of `With*` methods, a full set of `Add*` helpers (which are arguably are even simpler) but an incomplete set of `Remove*`.

> Was there a discussion on whether we need those new public APIs?
I wouldn't quite call it a discussion, but @CyrusNajmabadi took a look at API changes and agreed that these seemed like unintentional omissions.

jcouv replied:
Thanks for the clarifications.

lorcanmooney commented:
@jcouv  I've pushed a change with the additional test (see SemanticModelGetDeclaredSymbolAPITests.cs).

Do the public API changes require further approval, or is this good to merge?

jcouv commented:
LGTM Thanks

Get Outlook<https://aka.ms/uzin6x>

________________________________
From: Lorcan Mooney <notifications@github.com>
Sent: Friday, July 8, 2016 3:34:27 PM
To: dotnet/roslyn
Cc: Julien Couvreur; Mention
Subject: Re: [dotnet/roslyn] Add partial-type-name suggestions to C# and some fixes for VB (#12092)


@jcouv<https://github.com/jcouv> I've pushed a change with the additional test (see SemanticModelGetDeclaredSymbolAPITests.cs).

Do the public API changes require further approval, or is this good to merge?

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/dotnet/roslyn/pull/12092#issuecomment-231488689>, or mute the thread<https://github.com/notifications/unsubscribe/AL44Obd-lZoJO-SMJ7WQNiyqXcoEi4wXks5qTtBzgaJpZM4I49rt>.

lorcanmooney commented:
@Pilchie @CyrusNajmabadi Can this be merged now, or is there more to be done? I don't see any VSI test passes. Are they needed for this type of change?

Pilchie commented:
I'm not sure the current state of our VSI tests, but it's worth a shot, so test vsi please.

Pilchie commented:
Once those are done, I'm happy to merge this.

CyrusNajmabadi commented:
Right now only 'AnalyzersNode' seems to be flakey.  The rest of hte VSI tests seem ok.

tannergooding commented:
test vsi please. All tests failed due to ibc merge issues.

Pilchie commented:
All `ibcmerge` failures, which I think @tannergooding was looking at?

tannergooding commented:
Yes, I haved tagged @KevinH-MS to assist with the issue.

lorcanmooney commented:
Maybe today will by my lucky day: test vsi please

lorcanmooney commented:
*sighs* I guess not.

Can somebody let me know if the VSI failure is something I need to investigate?

jmarolf commented:
@lorcanmooney test failures are unrelated to your change.  Merging...
